relay-config/src/redis.rs

```diff
@@ -128,12 +130,90 @@ impl From<RedisConfigFromFile> for RedisConfig {
-    {
-        RedisConfigFromFile::deserialize(deserializer).map(Into::into)
+/// Configurations for the various Redis pools used by Relay.
+#[derive(Clone, Debug, Deserialize, Serialize, Eq, PartialEq)]
+#[serde(untagged)]
+pub enum RedisConfigs {
+    /// All pools should be configured the same way.
+    Unified(RedisConfig),
+    /// Individual configurations for each pool.
+    Individual {
+        /// Configuration for the `project_configs` pool.
+        project_configs: Box<RedisConfig>,
+        /// Configuration for the `cardinality` pool.
+        cardinality: Box<RedisConfig>,
+        /// Configuration for the `quotas` pool.
+        quotas: Box<RedisConfig>,
+        /// Configuration for the `misc` pool.
+        misc: Box<RedisConfig>,
+    },
+}
+
+/// Helper struct bundling connections and options for the various Redis pools.
+#[derive(Clone, Debug)]
+pub struct RedisPoolConfigs<'a> {
+    /// Configuration for the `project_configs` pool.
+    pub project_configs: (&'a RedisConnection, RedisConfigOptions),
+    /// Configuration for the `cardinality` pool.
+    pub cardinality: (&'a RedisConnection, RedisConfigOptions),
+    /// Configuration for the `quotas` pool.
+    pub quotas: (&'a RedisConnection, RedisConfigOptions),
+    /// Configuration for the `misc` pool.
+    pub misc: (&'a RedisConnection, RedisConfigOptions),
+}
+
+pub(super) fn create_redis_pool(
+    config: &RedisConfig,
+    cpu_concurrency: usize,
+) -> (&RedisConnection, RedisConfigOptions) {
+    let options = RedisConfigOptions {
+        max_connections: config
+            .options
+            .max_connections
+            .unwrap_or(cpu_concurrency as u32 * 2)
+            .min(crate::redis::DEFAULT_MIN_MAX_CONNECTIONS),
+        connection_timeout: config.options.connection_timeout,
+        max_lifetime: config.options.max_lifetime,
+        idle_timeout: config.options.idle_timeout,
+        read_timeout: config.options.read_timeout,
+        write_timeout: config.options.write_timeout,
+    };
+
+    (&config.connection, options)
+}
+
+pub(super) fn create_redis_pools(
+    configs: &RedisConfigs,
+    cpu_concurrency: usize,
+) -> RedisPoolConfigs {
+    match configs {
+        RedisConfigs::Unified(cfg) => {
+            let pool = create_redis_pool(cfg, cpu_concurrency);
+            RedisPoolConfigs {
+                project_configs: pool.clone(),
+                cardinality: pool.clone(),
+                quotas: pool.clone(),
+                misc: pool,
+            }
+        }
+        RedisConfigs::Individual {
+            project_configs,
+            cardinality,
+            quotas,
+            misc,
+        } => {
+            let project_configs = create_redis_pool(project_configs, cpu_concurrency);
+            let cardinality = create_redis_pool(cardinality, cpu_concurrency);
+            let quotas = create_redis_pool(quotas, cpu_concurrency);
+            let misc = create_redis_pool(misc, cpu_concurrency);
```

Comment:
Sorry, can you rephrase that? What should the `default_connections be for each of these pools? `min(2 * cpu_concurrency, DEFAULT_MIN_MAX_CONNECTIONS)` for the `project_configs` pool, just `2 * cpu_concurrency` for the rest? Or am I misunderstanding?

Reply:
Assuming `create_redis_pools` takes the default number of connections (instead of concurrency):

```rs
            let project_configs = create_redis_pool(project_configs, (cpu_concurrency * 2).min(crate::redis::DEFAULT_MIN_MAX_CONNECTIONS));
            let cardinality = create_redis_pool(cardinality, cpu_concurrency);
            let quotas = create_redis_pool(quotas, cpu_concurrency);
            let misc = create_redis_pool(misc, cpu_concurrency);
```

Reply:
It should be noted that the original code was
```rust
max_connections: redis
        .options
        .max_connections
        .unwrap_or(cpu_concurrency as u32 * 2)
        .min(crate::redis::DEFAULT_MIN_MAX_CONNECTIONS),
```
which does something different (it always clamps the connections to `DEFAULT_MIN_MAX_CONNECTIONS`, not just when `max_connections` isn't set).